### ISSUE TITLE
[Feature request]: `{:only}` on assertions

### ISSUE THREAD

**Issue by MikaelMayer** [dafny-lang/dafny#3095]
Dafny developers writing long methods often face long development cycles while developing the proofs of individual assertions. Even if they add `{:vcs_split_on_every_assert}`, the verifier uses wastes cores trying to verify things that are not relevant yet for the user.

As a workaround, It became almost idiomatic for Dafny developers to add `assume false;` to branches not in focus to speed-up verification.  This is because the verifier can use `false` to prove anything, so branches starting with `assume false;` are verified almost instantly, giving time for Dafny to focus on branches not having it. This mechanism is used both at the beginning of a proof (so that Dafny does not complain about multiple times the same unproven result) and during the proof (to "comment out" portions of the code that were already proven and that are not supposed to be changed)

However, consider the following example:

```
method Test(inputs...) {
  BigPreComputations(); // [1]
  match X {
  case 1 => assume false; // Proof already done, "commented out" by assume false;
  case 2 => // code in focus
  case 3 => assume false; // Proof to do...
  ...
  case n=> assume false; ...
 
  }
}
```

This emergent mechanism has major shortcomings:
- It does not scale with the number of branches. Focusing on one branches among n requires adding n-1 `assume false`
- All the assertions before the case split (in `[1]`) can't be marked with "assume false" easily; explicit assertions could using `by { assume false; }`, but implicit assertions would need to be made explicit, which, although easy in the case of preconditions, is not trivial for other kinds of assertions such as allocations or existentials.
- Forgetting an `assume false` somewhere could lead to a disastrous experience down the road when the compiler steps in. This could be solved by #3076 though.
- Sometimes there are so many facts available that I have personally seen a case where "assume false" did not speed up the verifier that much because it did not use that fact.

What developers really want is a way to tell the verifier to temporarily _focus_ on a part of the proof.
The good news, we already have experience of focusing. It works already for methods and functions.
* One can use `/proc:<functionName>` on the command line
* When verifying one function or method, all the others are accessed only by their axioms, and not their implementations.

Therefore, there is a straightforward mechanism to focus on a proof: we can temporarily transform assertions in assumptions, like we did when verifying a function, when we transform all the assertions of other functions into assumptions (axioms). This has the following advantages:
- Easy scaling: No more need to `assume false` in other branches, only one command or annotation per thing in focus.
- Ultra-fast temporary verification: Users would be able to iterate faster in sub-proofs by not having to worry that Dafny will spend substantial time trying to prove things that were already proven.
- Assuming a suitable UI like for #3076, as soon as the branch is proven, it would become obvious that the sub-goal has been reached. The compiler would also detect such constructs and fail to compile if they are used, akin to `assume false;`
- If multiple assertions takes time to verify, time could be dedicated for each individual assertions to optimize them, by simply hovering on the assertion every time it is verified, and looking at the resource count. This is very hard to achieve otherwise.

Now, how would users access this feature? Here are some possible user experiences, for each kind of user

### Command line

In the command line, we could have a command similar to `/proc` to indicate not the name of a method, but the position of an assertion to focus on (and all its proof in the "by" construct). It could take a name if the assertion is labelled, or a line number
Option name possibilities:
* `--procline`
* `--onlyVerifyLine`
* `--line`
* `--àssertOnlyLine`
* `--assertionName`
I've seen users actually use `/proc:` and an equivalent of `/trace` and then manually `grep` the results to focus only on one assertion. That would save them the trouble AND speed up their local verification.

### Attribute

We could introduce a new attribute, e.g. `{:assume-others}` (credits to @mattmccutchen-amazon for the name) . Alternatives are overriding the under-documented `{:focus}`, or another `{:only}`, `{:alone}`, `{:solo}` (like in music)... This proposal will go with the first suggested.
`{:assume-others}` could be used either independently like `assert {:assume-others} true;`, or on an assert ... by statement, e.g. like this

```
method Test(inputs...) ensures PostCondition {
  BigPreComputations(); // [1]
  match C {
  case 1 => // Proof already done here
  case 2 => assert {:assume-others} true; ... 
  case 2.5 => .... assert {:assume-others} Postcondition by { ... }
  case 3 => // Proof to do...
  ...
  case n=> ..
 
  }
}
```
I suggest the semantics to be the following
* `{:assume-others}` works at the level of one declaration. Because the verifier uses multiple cores and always reverifies the last method edited first, it's serves little purpose to deactivate all the other assertions in unrelated declarations by default. However, I also suggest below a way to do this, which can be useful when the verifier has a hard time canceling verification.
* `{:assume-others}` on an assertion without `by {...}` would make the assertion to be verified by assuming everything else.
* `{:assume-others}` on an assertion with `by {...}` would make the assertion and its proof to be isolated, that is, the proof will be also verified and not converted to assumptions.
* `{:assume-others}` on an assertion inside a `by {...}`  statement would "deactivate" any surrounding `{:assume-others}`. This is one way to temporarily focus on a sub-part of an already focused proof without having to change the parent's annotation
* `{:assume-others}` is the equivalent of `{:X 0}`
* `{:assume-others 1}` would isolate the surrounding block
* `{:assume-others 2}` would isolate the surrounding block's parent, and so on until the entire file is isolated, which is equivalent to no attribute (and we can warn the user about that)
* Two unrelated assertions marked with `{:assume-others}`  will both be verified. It's a way to selectively turn verification on for certain assertions, and assuming all the others. The dual of `assume false`, in some sense.

### IDE (e.g. VSCode) using the attribute

Before adding the attribute, all the assertions are verified on the left
![image](https://user-images.githubusercontent.com/3601079/203415630-8e465508-085d-43a0-82ed-fa403b1c9510.png)
After adding the attribute (mock image if the attribute's name is `{:X}`), the code is the same but only assertions in scope of the attribute are verified.
![image](https://user-images.githubusercontent.com/3601079/203415936-78307c4f-3d67-4ed3-9d99-2213c6750ff8.png)

### IDE (e.g. VSCode), automated

* (optional) A code action could suggest transforming a failing `assert A;` into `assert A by { <newline> }`;
* **Automatic focus:** As soon as the user edits a proof in a `by { ... }` where the assertion is failing, internally the language server could implicitly add the attribute `{:assume-others}` to the surrounding assertion. The verification would immediately visibly be in focused mode, and verification would be super fast.
* **Automatic de-focus:** If all the assertions now verify, the verifier would publish the partial results, automatically remove the implicit attribute `{:X}`, and relaunch the verification on the context.
* **Manual de-focus:** Another way to escape partial verification is that soon as the _caret_ moves out of the `by { ... }`, the verification would start immediately on the context as usual.
* **Manual focus:** When the user moves the caret in a `by { ... }`, in the test view, the IDE could display below the test of the method a "test" to represent the current proof, like this:
![image](https://user-images.githubusercontent.com/3601079/203418170-3749f9f5-c36f-4d3a-8e0e-518868e6c5a2.png)
Clicking or launching this test could trigger a focus so that the user does not need to remember any weird annotation `{:assume-others}`.

### Alternatives

There is no current viable alternative to this proposal besides heavily refactoring codebases and introducing smaller functions.
On the side, it was often request that one could run one single assertion at a time. But this has the drawback that when writing a proof, there are multiple assertions, we don't want to verify only one of them, but the entire proof. Also, the job to be done is usually proving a single assertion while assuming all the others. Moreover, as the number of assertions easily exceeds the number of lines of a function, it can become quite cumbersome to try to locate the right assertion. Clicking on the text editor and then viewing this single assertion as a single test would make it an easier interface.

**Comment by keyboardDrummer** [comment]
I like the automatic focus idea. I think it could work so that the IDE will always verify everything, but it'll automatically verify the block that you are editing first.

**Comment by MikaelMayer** [comment]
Thanks for explaining some alternatives. Let me comment on them for the record.

> Boogie has a system for verification caching at the statement level, where I believe it will automatically turn asserts into assumes if it knows they were verified before. Using this might allow us to get similar performance benefits as this proposal is hoping for, but with a faster time to verify everything.

I would be curious to see how it works. How does you detect if a simple assertion did not change?
Moreover, if your code is like this:
```
assert X; // not yet verified
assert Y by {
  // want to focus here
}
```
because X is not verified Boogie will try to verify it every time, whereas the approach of this feature proposal would actually be to just tell Boogie to ignore the assertion.

> Another alternative would be making it easier for users to refactor their codebase, such as providing an extract function extract method.

That would surely be valuable, but this is not an expected solution for the problem of the proof focus.
And that wouldn't work in many cases. Refactoring a function with complex types is already difficult, but refactoring a portion of some code by ensuring we have the right requirements is a nightmare. We don't have any mechanism yet to figure out what facts an assertion already depends on. 
Moreover, there are many cases where we don't want to refactor. If there are only two assertions in a method, and the second one is not passing, there is nothing to refactor.

**Comment by keyboardDrummer** [comment]
> if your code is like this:

```dafny
assert X; // not yet verified
assert Y by {
  // want to focus here
}
```

In that case I expect the user to want to make `assert X` verify before working on `assert Y`.

**Comment by MikaelMayer** [comment]
> > if your code is like this:
> 
> ```
> assert X; // not yet verified
> assert Y by {
>   // want to focus here
> }
> ```
> 
> In that case I expect the user to want to make `assert X` verify before working on `assert Y`.

That's a very high expectation. It's not uncommon in Dafny to require to prove two temporary assertions, and to focus on the second one. A user might not want to put assumptions because they eventually like to prove it anyway.

**Comment by seebees** [comment]
First this is great.
This feels like it would significantly simplify the process of verification debugging.

I note that you do not explicitly call out the problem of optimizing a proof.
Sometimes the proof exists,
but is not flexible enough.
I.e. it times out or fails in odd cases.
In this case everything has been proven,
I just want to figure out how to lead Dafny
more clearly to the solution.

To do this I need to be able to see the "cost"
of every proof obligation
so that I can pick where to spend my time. 

> but it'll automatically verify the block that you are editing first.

I also love this idea.
But I think that getting this to work perfectly is complicated and expensive.
By making a manual switch it is harder for strange interactions to pop up.
While the plugin is _much_ more stable,
I still see it crash or hang from time to time.

Moreover, as @MikaelMayer says I may want to focus
the second assertion first
because of my mental model of the problem.

> Another alternative would be making it easier for users to refactor their codebase

This can get very complicated.
To refactor out the actual logic
does not feel like a long term complete solution.
It does seem like a nice to have though.
Also, it makes me wonder about refactoring out
only the proof obligation.
With Dafny I have often written a lemma
to hoist out something that I need to prove.
That is another possibility.

But at the end of the day,
I think even that is not complete
because there is so much information
that I think the best experience
is to be able to focus on the specific location.

If we had such a focus feature,
I could see how this kind of refactor support
would be a powerful adjunct.
I could see wanting to do such a thing
without any proof obligation,
just because I want to refactor some logic
into a more portable function! :)

**Comment by keyboardDrummer** [comment]
> A user might not want to put assumptions because they eventually like to prove it anyway.

If they place an assumption without {:axiom} then Dafny should warn about that, so the user won't forget to change it.

**Comment by MikaelMayer** [comment]
> If they place an assumption without {:axiom} then Dafny should warn about that, so the user won't forget to change it.

Let's consider a bigger picture:

```
var x := f(y); // Precondition fails
assert P(x) by {
  ...
}
```
Now there is no assertion to transform to assumption. Will you accept that example?

**Comment by cpitclaudel** [comment]
@seebees off-topic for this thread, but why do you format your responses as poetry?

**Comment by keyboardDrummer** [comment]
> > If they place an assumption without {:axiom} then Dafny should warn about that, so the user won't forget to change it.
> 
> Let's consider a bigger picture:
> 
> ```
> var x := f(y); // Precondition fails
> assert P(x) by {
>   ...
> }
> ```
> 
> Now there is no assertion to transform to assumption. Will you accept that example?

I will, but I still find it strange that a user would want to skip a failing assertion and continue working on one that occurs later in the statement block. That's not how I think I program.

**Comment by robin-aws** [comment]
> @seebees off-topic for this thread, but why do you format your responses as poetry?

@cpitclaudel I would describe it as a zen-like devotion to https://sembr.org/ (unless @seebees would like to correct me - appealing to [Cunningham's Law](https://meta.wikimedia.org/wiki/Cunningham%27s_Law#:~:text=Cunningham's%20Law%20states%20%22the%20best,the%20inventor%20of%20wiki%20software.) here :)

**Comment by robin-aws** [comment]
What about providing a way to disable verification of individual assertion (batches) in the IDE, without changing the code? You could imagine listing all assertions and letting the user uncheck a single one, or uncheck all but one (which is the case you're focussing on here)? You could show that state in the gutter with a greyed-out icon.

This would be a much more direct equivalent feature to some kind of `--filter` CLI option (see the discussion on https://github.com/dafny-lang/dafny/pull/3051#discussion_r1028644090).

**Comment by MikaelMayer** [comment]
> What about providing a way to disable verification of individual assertion (batches) in the IDE, without changing the code? You could imagine listing all assertions and letting the user uncheck a single one, or uncheck all but one (which is the case you're focussing on here)? You could show that state in the gutter with a greyed-out icon.

That's a great idea. Let me comment it for the record as per why I think this great idea is not only complex to implement, but also has its own unresolved challenges, and finally why the feature request above should be implemented first before any other complex idea.

* In a single `assert`, there can be easily hundreds of individual assertions in industrial code, mainly because predicates are unfolded, and quantifiers are split. So we can't expect the user to find a single "assert" and check or uncheck it in a separate list. assert statements themselves are not visible in any assertion list we could produce today (they don't even have the same token), but this would be the adequate granularity to enable/disable something.
* Assertion batches were invented as part of Boogie, are documented in Dafny, but were never designed to help with proof development cycle in mind. To the contrary, I found that often, users rely on them to ensure less timeout rather than iterating on the proof to make it faster. Hence, my current (happy to debate it) opinion is that assertion batches save the day for some cases (especially with `{:vcs_split_on_every_assert}`), but don't solve the day to day job of letting users temporarily focus on certain parts of the proof to optimize them or just to make them verify. So far, only `assume false` did the job to most advanced Dafny users.

* Assertion batches are not named, and can overlap (a single assertion can belong to two different assertion batches, if you define a postcondition, it will belong to the last batch so visually two batches are on top of each other). This is confusing to think about in many cases. In this feature request, it's visually obvious what we would focus on, a single assertion at a time, or assertions in larger scopes. No need to think of complex assertion batches for that.
* When you click on the code, the test view is not like the outline and it does not know which test you are in. So you're left with trying to find the enclosing's method name, manually typing this name in the filter box) just to access the method to filter.
* Since it's not a standard component for a language server, it will have to be designed ad-hoc, and repeated in every IDE out there.
* The current way we display symbols and their verifications (a.k.a. the test view) does not allow for more affordances, such as filtering which symbols to verify when editing the document in real-time. We would have to reimplement everything in a Webview if we want to customize this.
* It is almost impossible to track where assertions are moving to. So if a user cut and paste a proof they wants to keep focus on, they have to do all the steps again:
  * Splitting in batches by writing `assert {:split_here} true` twice in their code
  * Looking for the batch in the test view which would have been converted to a webview to ensure we can filter things
  * Locating the batch among all the unnamed ones after the first verification (it does not appear before currently, or we would need to develop this)
  * Right-clicking or clicking on a button to only verify that batch, and then going back to the proof.
* If a proof takes time and the user has to quit VSCode, the IDE state would be lost and they would have to do this again. By recognizing the need as an attribute, this attribute could be kept from one session to the next.
* Once the proof is straightforward, the user will have to manually remove the two splits introduced before and after.

Now compare all these unresolved to the simplicity of the feature request above:

> A user writes {:assume-others} on an assert.

And the potential implementation:

> In Dafny, when encountering `{:assume-others}`, we could add `by { assume false; ... rest of proof}` after every other assertion not in scope of the assertion with `{:assume-others}`.

I have been using this strategy today to fix the problem of a now happy Dafny customer, to isolate the problematic assertion without having to isolate its context. The strategy above was the only viable one because some assertions were named, and we cannot name assumptions (but that could also be another way to implement this feature)

We can even put a lightbulb on any failing assert to let users discover/remove this attribute.

Now, how your idea could become true in the future:
* Once we have this attribute, we could enable/disable programmatically instead of in the code, depending on some UI affordance we will have agreed on, such as a list view with filter.

I hope I have convinced you of the challenges related to your proposal so that you better see how this feature request is a precursor of what you have in mind.

**Comment by MikaelMayer** [comment]
> > if your code is like this:
> 
> ```
> assert X; // not yet verified
> assert Y by {
>   // want to focus here
> }
> ```
> 
> In that case I expect the user to want to make `assert X` verify before working on `assert Y`.

This morning I ran into an example that does not match your expectations, let me walk you through it.

Basically, X is a precondition for Y, and Y contains a call to a function F whose body was not yet written for the branch I'm considering. To prove Y, I first needed to write the body of F - meaning I knew the spec, but not the body.
I can only derive the body from the Y. Indeed, X is too weak. Hence, even though I wish I could prove X first, the two assertions would actually be proven at the same time normaly. I found myself focusing again on the "by" clause of the `assert Y` until I found the correct body of a function, thanks to Y.
After I was done, X was automatically verified because it also depended on the body of F
And meanwhile, to ensure my verification process in the same function wasn't polluted by other errors like on X which I know would go away later, I had to manually replace "assert X" by "assume X", and then back to "assert X" when I was done. But really, I was just focusing my proof on Y, if with a code action I could have said "now, assume all the preconditions" (because there were more than X), I would have preferred that.
